src/client/datascience/interactive-window/nativeInteractiveWindowProvider.ts

```diff
@@ -0,0 +1,227 @@
+    }
+    public get onDidCreateInteractiveWindow(): Event<IInteractiveWindow> {
+        return this._onDidCreateInteractiveWindow.event;
+    }
+    public get activeWindow(): IInteractiveWindow | undefined {
+        return this._windows.find((w) => w.active && w.visible);
+    }
+    public get windows(): ReadonlyArray<IInteractiveWindow> {
+        return this._windows;
+    }
+    private readonly _onDidChangeActiveInteractiveWindow = new EventEmitter<IInteractiveWindow | undefined>();
+    private readonly _onDidCreateInteractiveWindow = new EventEmitter<IInteractiveWindow>();
+    private lastActiveInteractiveWindow: IInteractiveWindow | undefined;
+    private _windows: NativeInteractiveWindow[] = [];
+    constructor(
+        @inject(IServiceContainer) private serviceContainer: IServiceContainer,
+        @inject(IAsyncDisposableRegistry) asyncRegistry: IAsyncDisposableRegistry,
+        @inject(IDisposableRegistry) private disposables: IDisposableRegistry,
+        @inject(IFileSystem) private readonly fs: IFileSystem,
+        @inject(IConfigurationService) private readonly configService: IConfigurationService,
+        @inject(IMemento) @named(GLOBAL_MEMENTO) private readonly globalMemento: Memento,
+        @inject(IApplicationShell) private readonly appShell: IApplicationShell,
+        @inject(IWorkspaceService) private readonly workspace: IWorkspaceService,
+        @inject(IKernelProvider) private readonly kernelProvider: IKernelProvider,
+        @inject(INotebookControllerManager) private readonly notebookControllerManager: INotebookControllerManager
+    ) {
+        asyncRegistry.push(this);
+    }
+
+    public async getOrCreate(resource: Resource): Promise<IInteractiveWindow> {
+        if (!this.workspace.isTrusted) {
+            // This should not happen, but if it does, then just throw an error.
+            // The commands the like should be disabled.
+            throw new Error('Worksapce not trusted');
+        }
+        // Ask for a configuration change if appropriate
+        const mode = await this.getInteractiveMode(resource);
+
+        // See if we already have a match
+        let result = this.get(resource, mode) as IInteractiveWindow;
+        if (!result) {
+            // No match. Create a new item.
+            result = await this.create(resource, mode, undefined);
+        }
+
+        return result;
+    }
+
+    public async dispose(): Promise<void> {
+        return noop();
+    }
+
+    public async synchronize(_window: IInteractiveWindow): Promise<void> {
+        // TODO delete this method entirely
+        noop();
+    }
+
+    protected async create(
+        resource: Resource,
+        mode: InteractiveWindowMode,
+        notebookUri: Uri | undefined
+    ): Promise<NativeInteractiveWindow> {
+        // Ensure all our controllers are registered with VS Code
+        await this.notebookControllerManager.loadNotebookControllers();
```

Comment:
> Is this so that we can execute a cell as soon as we open the interactive window?

Yes, we want controllers to be available ASAP for the interactive window so that we can immediately execute code that's submitted from a #%%-interactive file